internal/controller/pkg/signature/validate.go

```diff
@@ -0,0 +1,197 @@
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/sigstore/cosign/v2/pkg/cosign"
+	ociremote "github.com/sigstore/cosign/v2/pkg/oci/remote"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sigstore/sigstore/pkg/fulcioroots"
+	"github.com/sigstore/sigstore/pkg/signature"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+
+	"github.com/crossplane/crossplane/apis/pkg/v1beta1"
+)
+
+const fetchCertTimeout = 30 * time.Second
+
+// Validator validates image signatures.
+type Validator interface {
+	Validate(ctx context.Context, ref name.Reference, config *v1beta1.ImageVerification, opts ...remote.Option) error
+}
+
+// NewCosignValidator returns a new CosignValidator.
+func NewCosignValidator(c client.Reader, namespace string) (*CosignValidator, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), fetchCertTimeout)
+	defer cancel()
+
+	var err error
+	opts := &cosign.CheckOpts{}
+	opts.RootCerts, err = fulcioroots.Get()
+	if err != nil {
+		return nil, errors.Errorf("cannot fetch Fulcio roots: %w", err)
+	}
+	opts.IntermediateCerts, err = fulcioroots.GetIntermediates()
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch Fulcio intermediates: %w", err)
+	}
+	opts.CTLogPubKeys, err = cosign.GetCTLogPubs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch CTLog public keys: %w", err)
+	}
+
+	opts.RekorPubKeys, err = cosign.GetRekorPubs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch Rekor public keys: %w", err)
+	}
+
+	return &CosignValidator{
+		client:    c,
+		namespace: namespace,
+
+		checkOpts: opts,
+	}, nil
+}
+
+// CosignValidator validates image signatures using cosign.
+type CosignValidator struct {
+	client    client.Reader
+	namespace string
+
+	checkOpts *cosign.CheckOpts
+}
+
+// Validate validates the image signature.
+func (c *CosignValidator) Validate(ctx context.Context, ref name.Reference, config *v1beta1.ImageVerification, opts ...remote.Option) error {
+	if config.Provider != v1beta1.ImageVerificationProviderCosign {
+		return errors.New("unsupported image verification provider")
+	}
+
+	var errs []error
+	for _, a := range config.Cosign.Authorities {
+		c.checkOpts.ClaimVerifier = cosign.SimpleClaimVerifier
+		verify := cosign.VerifyImageSignatures
+		if len(a.Attestations) > 0 {
+			c.checkOpts.ClaimVerifier = cosign.IntotoSubjectClaimVerifier
+			verify = cosign.VerifyImageAttestations
+		}
+
+		if err := c.buildCosignCheckOpts(ctx, a, ociremote.WithRemoteOptions(opts...)); err != nil {
+			errs = append(errs, errors.Errorf("authority %q: cannot build cosign check options %v", a.Name, err))
+			continue
+		}
+
+		res, ok, err := verify(ctx, ref, c.checkOpts)
+		if err != nil {
+			errs = append(errs, errors.Errorf("authority %q: signature verification failed with %v", a.Name, err))
+			continue
+		}
```

Comment:
Is there a reason this seems to mutate the validator scoped check opts on each iteration?

It seems like there's a couple of validator scoped check opts (e.g. those setup in `NewCosignValidator`), plus some for each authority (setup by `c.buildCosignCheckOpts`). Could `c.buildCosignCheckOpts` return the options  for each authority (presumably a superset of those setup by `NewCosignValidator`) instead?

Reply:
Good call, done!